modules/tensor_mechanics/src/materials/abaqus/AbaqusUmatMaterial.C

```diff
@@ -0,0 +1,262 @@
+    _stress_old(getMaterialPropertyOld<RankTwoTensor>(_base_name + "stress")),
+    _total_strain(getMaterialProperty<RankTwoTensor>(_base_name + "total_strain")),
+    _strain_increment(getMaterialProperty<RankTwoTensor>(_base_name + "strain_increment")),
+    _jacobian_mult(declareProperty<RankFourTensor>(_base_name + "Jacobian_mult")),
+    _Fbar(getMaterialPropertyOld<RankTwoTensor>(_base_name + "deformation_gradient")),
+    _Fbar_old(getMaterialPropertyOld<RankTwoTensor>(_base_name + "deformation_gradient")),
+    _state_var(declareProperty<std::vector<Real>>(_base_name + "state_var")),
+    _state_var_old(getMaterialPropertyOld<std::vector<Real>>(_base_name + "state_var")),
+    _elastic_strain_energy(declareProperty<Real>(_base_name + "elastic_strain_energy")),
+    _plastic_dissipation(declareProperty<Real>(_base_name + "plastic_dissipation")),
+    _creep_dissipation(declareProperty<Real>(_base_name + "creep_dissipation")),
+    _material_timestep(declareProperty<Real>(_base_name + "material_timestep_limit"))
+{
+#if defined(METHOD)
+  _plugin += std::string("-") + QUOTE(METHOD) + ".plugin";
+#endif
+
+  // Size and create full (mechanical+thermal) material property array
+  _num_props = _mechanical_constants.size() + _thermal_constants.size();
+
+  // Read mesh dimension and size UMAT arrays (we always size for full 3D)
+  _aqNTENS = 6; // Size of the stress or strain component array (NDI+NSHR)
+  _aqNSHR = 3;  // Number of engineering shear stress components
+  _aqNDI = 3;   // Number of direct stress components (always 3)
+
+  _aqSTATEV.resize(_num_state_vars);
+  _aqDDSDDT.resize(_aqNTENS);
+  _aqDRPLDE.resize(_aqNTENS);
+  _aqSTRAN.resize(_aqNTENS);
+  _aqDFGRD0.resize(9);
+  _aqDFGRD1.resize(9);
+  _aqSTRESS.resize(_aqNTENS);
+  _aqDDSDDE.resize(_aqNTENS * _aqNTENS);
+  _aqDSTRAN.resize(_aqNTENS);
+  _aqPROPS.resize(_num_props);
+
+  // Assign materials properties from vector form into an array
+  for (unsigned int i = 0; i < _mechanical_constants.size(); ++i)
+    _aqPROPS[i] = _mechanical_constants[i];
+  for (unsigned int i = _mechanical_constants.size(); i < _num_props; ++i)
+    _aqPROPS[i] = _thermal_constants[i];
+
+  // Size UMAT state variable (NSTATV) and material constant (NPROPS) arrays
+  _aqNSTATV = _num_state_vars;
+  _aqNPROPS = _num_props;
+
+  // Open the library
+#ifdef LIBMESH_HAVE_DLOPEN
+  _handle = dlopen(_plugin.c_str(), RTLD_LAZY);
+
+  if (!_handle)
+    mooseError("Cannot open library: ", dlerror());
+
+  // Reset errors
+  dlerror();
+
+  // Snag the function pointer from the library
+  {
+    void * pointer = dlsym(_handle, "umat_");
+    _umat = *reinterpret_cast<umat_t *>(&pointer);
+  }
+
+  // Catch errors
+  const char * dlsym_error = dlerror();
+  if (dlsym_error)
+  {
+    dlclose(_handle);
+    std::ostringstream error;
+    error << "Cannot load symbol 'umat_': " << dlsym_error << '\n';
+    mooseError(error.str());
+  }
+#else
+  mooseError("AbaqusUmatMaterial is not supported on Windows.");
```

Comment:
Is Windows the only platform where `LIBMESH_HAVE_DLOPEN` is false? Maybe that error should be made more general.